libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/LocationEngineWrapper.kt

```diff
@@ -0,0 +1,79 @@
+import com.mapbox.android.core.location.LocationEngineCallback
+import com.mapbox.android.core.location.LocationEngineRequest
+import com.mapbox.android.core.location.LocationEngineResult
+import java.lang.Exception
+
+/**
+ * LocationEngine does not have a method to cancel get last location task,
+ * as it has `removeLocationUpdates` for `requestLocationUpdates`.
+ * This may lead to getLastLocation callback firing after `stopLocationUpdates` was invoked,
+ * which is not expected behaviour in context of Mapbox trip sessions.
+ * This class introduces an ability to cancel get last location task for a particular callback.
+ * The issue is reproducible with `ReplayLocationTest#replay_session_locations_do_not_contain_locations_from_previous_session`.
+ */
+internal class LocationEngineWrapper(
+    private val originalEngine: LocationEngine
+) : LocationEngine {
+
+    private val lastLocationCallbacks =
+        mutableListOf<LocationEngineCallback<LocationEngineResult>>()
+
+    @SuppressLint("MissingPermission")
+    override fun getLastLocation(callback: LocationEngineCallback<LocationEngineResult>) {
+        lastLocationCallbacks.add(callback)
+        originalEngine.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {
+
+            override fun onSuccess(result: LocationEngineResult?) {
+                if (lastLocationCallbacks.remove(callback)) {
+                    callback.onSuccess(result)
+                }
+            }
+
+            override fun onFailure(exception: Exception) {
+                if (lastLocationCallbacks.remove(callback)) {
+                    callback.onFailure(exception)
+                }
+            }
+        })
+    }
+
+    /**
+     * Cancel getLastLocation for a specified callback.
+     * If original engine triggers callback for getLastLocation task after this method is invoked,
+     * the update will not be proxied to the original callback.
+     */
+    fun cancelLastLocationTask(callback: LocationEngineCallback<LocationEngineResult>) {
+        lastLocationCallbacks.remove(callback)
+    }
+
+    @SuppressLint("MissingPermission")
+    override fun requestLocationUpdates(
+        request: LocationEngineRequest,
+        callback: LocationEngineCallback<LocationEngineResult>,
+        looper: Looper?
+    ) {
+        originalEngine.requestLocationUpdates(request, callback, looper)
+    }
+
+    @SuppressLint("MissingPermission")
+    override fun requestLocationUpdates(
+        request: LocationEngineRequest,
+        pendingIntent: PendingIntent?
+    ) {
+        originalEngine.requestLocationUpdates(request, pendingIntent)
+    }
+
+    override fun removeLocationUpdates(callback: LocationEngineCallback<LocationEngineResult>) {
+        originalEngine.removeLocationUpdates(callback)
+    }
+
+    override fun removeLocationUpdates(pendingIntent: PendingIntent?) {
+        originalEngine.removeLocationUpdates(pendingIntent)
+    }
+}
```

Comment:
Maybe this is good use case for the [delegation feature](https://kotlinlang.org/docs/delegation.html#overriding-a-member-of-an-interface-implemented-by-delegation)?
```suggestion
internal class LocationEngineWrapper(
    private val originalEngine: LocationEngine
) : LocationEngine by originalEngine {

    private val lastLocationCallbacks =
        mutableListOf<LocationEngineCallback<LocationEngineResult>>()

    @SuppressLint("MissingPermission")
    override fun getLastLocation(callback: LocationEngineCallback<LocationEngineResult>) {
        lastLocationCallbacks.add(callback)
        originalEngine.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {

            override fun onSuccess(result: LocationEngineResult?) {
                if (lastLocationCallbacks.remove(callback)) {
                    callback.onSuccess(result)
                }
            }

            override fun onFailure(exception: Exception) {
                if (lastLocationCallbacks.remove(callback)) {
                    callback.onFailure(exception)
                }
            }
        })
    }

    /**
     * Cancel getLastLocation for a specified callback.
     * If original engine triggers callback for getLastLocation task after this method is invoked,
     * the update will not be proxied to the original callback.
     */
    fun cancelLastLocationTask(callback: LocationEngineCallback<LocationEngineResult>) {
        lastLocationCallbacks.remove(callback)
    }
}
```

Reply:
Haven't really used this a lot so I didn't think of it.
Yep, looks like that's exactly what I need.